[JENKINS-47372] Add a new Administrative monitor for CSRF-protection

Wadeck opened the pull request:
Add administrative monitor when there is no CSRF issuer configured.

See [JENKINS-47372](https://issues.jenkins-ci.org/browse/JENKINS-47372).

### Proposed changelog entries

* Add new administrative monitor to ensure CSRF is enabled.

### Submitter checklist

- [x] JIRA issue is well described
- [x] Changelog entry appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
      * Use the `Internal: ` prefix if the change has no user-visible impact (API, test frameworks, etc.)
- [x] Appropriate autotests or explanation to why this change has no tests
- [x] For dependency updates: links to external changelogs and, if possible, full diffs

<!-- For new API and extension points: Link to the reference implementation in open-source (or example in Javadoc) -->

### Desired reviewers

@reviewbybees

<!-- Comment:
If you need an accelerated review process by the community (e.g., for critical bugs), mention @jenkinsci/code-reviewers
-->

Screenshot of the message in action:
![csrf_administrative_monitor_popup](https://user-images.githubusercontent.com/2662497/31395605-6788b172-ade1-11e7-9d99-b322eee3eb34.PNG)

Wadeck committed:
[JENKINS-47372] add administrative monitor when there is no CSRF issuer configured

oleg-nenashev commented on core/src/main/resources/jenkins/security/csrf/Messages.properties:
```diff
@@ -0,0 +1 @@
+CSRFAdministrativeMonitor.displayName=CSRF Protection Monitor
```
Nice2have: line break

oleg-nenashev commented on core/src/main/resources/jenkins/security/csrf/CSRFAdministrativeMonitor/message.properties:
```diff
@@ -0,0 +1,3 @@
+warningMessage=You have not configured the CSRF issuer. This could be a security issue. \
+  For more information, please refer to 
+changeConfiguration=You can change the current configuration using the Security section 
```
Nice2have: line break

oleg-nenashev commented on core/src/main/resources/jenkins/security/csrf/CSRFAdministrativeMonitor/message.jelly:
```diff
@@ -0,0 +1,10 @@
+        ${%changeConfiguration}
+        <a href="${rootURL}/configureSecurity">${%CSRF Protection}</a>.
+    </div>
+</j:jelly>
```
Nice2have: line break

oleg-nenashev reviewed (approved):
🐝

Wadeck committed:
- add line breaks

daniel-beck commented on core/src/main/java/jenkins/security/csrf/CSRFAdministrativeMonitor.java:
```diff
@@ -0,0 +1,37 @@
+ * Monitor that the CSRF is enabled on the application.
+ *
+ * @author Wadeck Follonier
+ * @author Oleg Nenashev
```
Please don't `@author`.

Wadeck replied:
removed as requested

daniel-beck commented on core/src/main/java/jenkins/security/csrf/CSRFAdministrativeMonitor.java:
```diff
@@ -0,0 +1,37 @@
+package jenkins.security.csrf;
```
Missing license header.

Wadeck replied:
added as requested

daniel-beck commented on core/src/main/java/jenkins/security/csrf/CSRFAdministrativeMonitor.java:
```diff
@@ -0,0 +1,37 @@
+ * @since TODO
+ */
+@Extension
+@Symbol("csrf")
```
Are we still maintaining these? Why? Isn't https://github.com/jenkinsci/system-config-dsl-plugin dead?

oleg-nenashev replied:
It's not dead. @ewelinawilkosz2 and @Andrey9kin adopted it.
https://github.com/Praqma/system-config-dsl-plugin

So I would keep symbol annotations when possible. CC @ndeloof

daniel-beck replied:
Oh, that's nice. I was unaware this is `system-config-dsl` based.

ndeloof replied:
Agree. We should have @symbol mostly everywhere on components which end user might need to access for configuration, as sort of a 'natural' namr

ewelinawilkosz2 replied:
Just to be clear, we're now working on https://github.com/Praqma/jenkins-coco-plugin which is `system-config-dsl-plugin` based but with a new way of doing some things. As soon as we decide this is how we want to keep things we'll try to announce it using known channels

daniel-beck commented on core/src/main/java/jenkins/security/csrf/CSRFAdministrativeMonitor.java:
```diff
@@ -0,0 +1,37 @@
+            return true;
+        }
+        
+        return false;
```
`return Jenkins.getInstance().getCrumbIssuer() == null`?

Wadeck replied:
simplified as proposed

daniel-beck commented on core/src/main/resources/jenkins/security/csrf/CSRFAdministrativeMonitor/message.jelly:
```diff
@@ -0,0 +1,10 @@
+<j:jelly xmlns:j="jelly:core">
+    <div class="warning">
+        ${%warningMessage}
+        <a href="https://jenkins.io/redirect/csrf-protection" target="_blank">${%this page}</a>.
```
This seems like a hackish combination of localized strings (not every language may be able to finish the sentence with what makes sense as a link label). Would be preferable to have the link inside the localized string.

oleg-nenashev replied:
It would require wrapping the localized string with `j:out` IIUC. I am not sure I am ready to advertise it

daniel-beck replied:
It won't. There's a specific behavior implemented in Stapler Jelly to not escape HTML from localized strings. So just move the entire `div` content in there, and done.

Wadeck replied:
corrected as requested, thanks for the information about trusted html in properties

daniel-beck commented on core/src/main/resources/jenkins/security/csrf/CSRFAdministrativeMonitor/message.jelly:
```diff
@@ -0,0 +1,10 @@
+<?jelly escape-by-default='true'?>
```
Missing license header.

Wadeck replied:
added as requested

daniel-beck commented on core/src/main/resources/jenkins/security/csrf/CSRFAdministrativeMonitor/message.jelly:
```diff
@@ -0,0 +1,10 @@
+<?jelly escape-by-default='true'?>
+<j:jelly xmlns:j="jelly:core">
```
Jelly 😢

daniel-beck commented on core/src/main/java/jenkins/security/csrf/CSRFAdministrativeMonitor.java:
```diff
@@ -0,0 +1,37 @@
+package jenkins.security.csrf;
```
Seems a bit much to create a separate package for this, no?

Notably, crumb issuers and related existing code cannot be moved anyway.

oleg-nenashev replied:
Usually we create mirror packages in Jenkins for the new things. I can imagine new classes around csrf things, so having a new package LGTM.

daniel-beck replied:
Ah, I missed that `hudson.security.csrf` exists. 👍 then

daniel-beck reviewed (changes requested):
Some minor findings, but the i18n making assumptions about sentence structure needs fixing.

reviewbybees commented:
This pull request originates from a [CloudBees](https://www.cloudbees.com/) employee. At CloudBees, we require that all pull requests be reviewed by other CloudBees employees before we seek to have the change accepted. If you want to learn more about our process please see [this explanation](https://github.com/reviewbybees/about#about-reviewbybees).

Wadeck committed:
- add license header
- put link in the properties instead of the previous mix
- remove @author
- simplify isActivated body

Wadeck committed:
- correct line breaks

recena commented:
This PR would look much better with this one https://github.com/jenkinsci/jenkins/pull/2857 😄

oleg-nenashev reviewed (approved):
re-:bee: . Would be great to have it in the next LTS baseline

daniel-beck reviewed (approved):
Assuming this has been manually tested as well, sure.

oleg-nenashev commented:
Ready to go, I'd guess
